src/msg/async/dpdk/EventDPDK.cc

```diff
@@ -25,61 +25,109 @@
 
 #undef dout_prefix
 #define dout_prefix *_dout << "DPDKDriver."
+#define EVENTS_MAX 64
 
 int DPDKDriver::init(EventCenter *c, int nevent)
 {
+	epfd = epoll_create(EVENTS_MAX);
+	if (epfd < 0) {
+            ldout(cct, 0) << __func__ << "can't create epoll fd" << dendl;
+	    return -EIO;
+	}
+	if (::fcntl(epfd, F_SETFD, FD_CLOEXEC) == -1) {
+	   int e = errno;
+	   ::close(epfd);
+	   lderr(cct) << __func__<< " unable to set cloexec: " << strerror(errno) << dendl;
+
+	   return e;
+	}
 	return 0;
 }
 
 int DPDKDriver::add_event(int fd, int cur_mask, int add_mask)
 {
 	ldout(cct, 20) << __func__ << " add event fd=" << fd << " cur_mask=" << cur_mask
-								 << " add_mask=" << add_mask << dendl;
-
-	int r = manager.listen(fd, add_mask);
-	if (r < 0) {
-		lderr(cct) << __func__ << " add fd=" << fd << " failed. "
-		           << cpp_strerror(-r) << dendl;
-		return -errno;
+		       << " add_mask=" << add_mask << dendl;
+        int r;
+	if (fd != wakeup_fd) {
+	   r = manager.listen(fd, add_mask);
+	} else {
+	    struct epoll_event ee;
+	    ee.events = EPOLLIN | EPOLLET;
+	    ee.data.u64 = 0;
+	    ee.data.fd = wakeup_fd;
+	    r = epoll_ctl(epfd, EPOLL_CTL_ADD, wakeup_fd, &ee);
+	}
+        if (r < 0) {
+	    lderr(cct) << __func__<< " add fd=" << fd << " failed. " << cpp_strerror(-r) << dendl;
+	    return -errno;
 	}
-
 	return 0;
 }
 
 int DPDKDriver::del_event(int fd, int cur_mask, int delmask)
 {
-	ldout(cct, 20) << __func__ << " del event fd=" << fd << " cur_mask=" << cur_mask
-								 << " delmask=" << delmask << dendl;
-	int r = 0;
+	ldout(cct, 20) << __func__ << " del event fd=" << fd << " cur_mask=" << cur_mask << " delmask=" << delmask << dendl;
+	if (delmask == EVENT_NONE)
+	return 0;
 
-	if (delmask != EVENT_NONE) {
-		if ((r = manager.unlisten(fd, delmask)) < 0) {
-			lderr(cct) << __func__ << " delete fd=" << fd << " delmask=" << delmask
-								 << " failed." << cpp_strerror(-r) << dendl;
-			return r;
-		}
+	int r;
+        if (fd != wakeup_fd) {
+	   r = manager.unlisten(fd, delmask);
+	} else {
+	   r = epoll_ctl(epfd, EPOLL_CTL_DEL, wakeup_fd, nullptr);
+	}
+	if (r < 0) {
+	   lderr(cct) << __func__ << " delete fd=" << fd << " delmask=" << delmask
+		      << " failed." << cpp_strerror(-r) << dendl;
+	   return r;
 	}
 	return 0;
 }
 
 int DPDKDriver::resize_events(int newsize)
 {
-	return 0;
+       return 0;
```

Comment:
if we are switching from <tab> to <spaces>, please be consistent at least in this very commit.